keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -0,0 +1,266 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Manage the vendor folder in `kubernetes/kubernetes` using go modules,
+and define `go.mod` module files for published components like `k8s.io/client-go` and `k8s.io/api`.
+
+## Motivation
+
+Since its inception, Kubernetes has used Godep to manage vendored 
+dependencies, to ensure reproducible builds and auditable source.
+
+As the go ecosystem matured, vendoring became a first-class concept,
+Godep became unmaintained, Kubernetes started using a custom version of Godep,
+other vendoring tools (like `glide` and `dep`) became available,
+and dependency management was ultimately added directly to go in the form of go modules.
+
+The [plan of record](https://blog.golang.org/modules2019) is for go1.13 to enable
+go modules by default and deprecate GOPATH mode. To be ready for that, Kubernetes
+components must make any required adjustments in the Kubernetes 1.15 release.
+
+In addition to simply keeping up with the go ecosystem, go modules provide many benefits:
+* rebuilding `vendor` with go modules provided a 10x speed increase over Godep in preliminary tests
+* go modules can reproduce a consistent `vendor` directory on any OS
+* if semantic import versioning is adopted, consumers of Kubernetes modules can use two distinct versions simultaneously (if required by diamond dependencies)
+
+### Goals
+
+* Remove the kubernetes/kubernetes dependency on godep (including the need for a customized version of godep)
+* Increase speed and reliability of building and verifying the `vendor` directory in kubernetes/kubernetes
+* Provide accurate go.mod module descriptors for kubernetes/kubernetes and published staging components
+* Enable use of published kubernetes components by `go module` aware consumers
+* Allow continued use of published kubernetes components by `go module` *unaware* consumers until go modules are enabled by default
+
+## Proposal
+
+### Manage vendor folders using go modules
+1. Make `k8s.io/kubernetes` and each staging component (like `k8s.io/client-go` and `k8s.io/api`) a distinct go module
+  
+    * Pin dependencies to the same versions that are currently listed in Godeps.json
+
+2. Change vendor management and verification scripts in `kubernetes/kubernetes` to use go module commands
+3. Change the staging component publishing bot to modify the `go.mod` pinned versions of peer components, rather than rewriting Godeps.json files
```

Comment:
I thought under the module system this publishing step wouldn't be necessary at all anymore, as a single repo can declare multiple modules?

Reply:
there are four primary challenges I found with switching (ordered from most to least fundamental):
* git shas from before 1.15 would not be available as go modules, so would need to continue being accessed via the individual repos. I couldn't find a way to split the location based on version.
* pointing to a relative location for a module like `k8s.io/api` [works well](https://github.com/liggitt/kubernetes/blob/go-mod-cleanup/go.mod#L28-L54), but it's unclear how a third-party would use a replace directive to locate the nested module
* there doesn't seem to be a way to point vanity imports at anything other than a repo root when pointing at a VCS
* the kubernetes/kubernetes repository is extremely large, and forcing a clone of it to pick up `k8s.io/api`, for example, is unpleasant

I see this proposal as a first step, with possible future simplification/removal of the publishing process if the above issues are resolved

Reply:
s/unpleasant/abusive/

As always you are a master of understatement.

Reply:
Yeah, I completely agree with not doing everything in one big step. It might be good to incorporate those reasons in the text.

Reply:
Will fold into next PR

Reply:
added in https://github.com/kubernetes/enhancements/pull/916